src/bin/pg_autoctl/pghba.c

```diff
@@ -36,76 +42,101 @@ static int escape_hba_string(char *destination, const char *hbaString);
-	PQExpBuffer newHbaContents = NULL;
-
-	if (hbaLineBuffer == NULL)
-	{
-		log_error("Failed to allocate memory");
-		return false;
-	}
-
 	if (ssl)
 	{
-		appendPQExpBufferStr(hbaLineBuffer, "hostssl ");
+		appendPQExpBufferStr(buffer, "hostssl ");
 	}
 	else
 	{
-		appendPQExpBufferStr(hbaLineBuffer, "host ");
+		appendPQExpBufferStr(buffer, "host ");
 	}
 
-	append_database_field(hbaLineBuffer, databaseType, database);
-	appendPQExpBufferStr(hbaLineBuffer, " ");
+	append_database_field(buffer, databaseType, database);
+	appendPQExpBufferStr(buffer, " ");
 
 	if (username)
 	{
 		char escapedUsername[BUFSIZE] = { 0 };
 		(void) escape_hba_string(escapedUsername, username);
 
-		appendPQExpBufferStr(hbaLineBuffer, escapedUsername);
-		appendPQExpBufferStr(hbaLineBuffer, " ");
+		appendPQExpBufferStr(buffer, escapedUsername);
+		appendPQExpBufferStr(buffer, " ");
 	}
 	else
 	{
-		appendPQExpBufferStr(hbaLineBuffer, "all ");
+		appendPQExpBufferStr(buffer, "all ");
 	}
 
-	append_hostname_or_cidr(hbaLineBuffer, host);
-	appendPQExpBuffer(hbaLineBuffer, " %s", authenticationScheme);
+	append_hostname_or_cidr(buffer, host);
+	appendPQExpBuffer(buffer, " %s", authenticationScheme);
 
-	if (PQExpBufferBroken(hbaLineBuffer))
+	if (PQExpBufferBroken(buffer))
 	{
 		log_error("Failed to allocate memory");
-		destroyPQExpBuffer(hbaLineBuffer);
+		destroyPQExpBuffer(buffer);
 		return false;
 	}
 
-	/*
-	 * When the authentication method is "skip", the option --skip-pg-hba has
-	 * been used. In that case, we still WARN about the HBA rule that we need,
-	 * so that users can review their HBA settings and provisioning.
-	 */
-	if (SKIP_HBA(authenticationScheme))
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(buffer))
 	{
-		log_warn("Skipping HBA edits (per --skip-pg-hba) for rule: %s",
-				 hbaLineBuffer->data);
-		return true;
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(buffer);
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * pghba_ensure_host_rule_exists ensures that a host rule exists in the
+ * pg_hba file with the given database, username, host and authentication
+ * scheme.
+ */
+bool
+pghba_ensure_host_rule_exists(const char *hbaFilePath,
+							  bool ssl,
+							  HBADatabaseType databaseType,
+							  const char *database,
+							  const char *username,
+							  const char *host,
+							  const char *authenticationScheme)
+{
+	char *currentHbaContents = NULL;
+	long currentHbaSize = 0L;
+	char *includeLine = NULL;
+	PQExpBuffer hbaLineBuffer = createPQExpBuffer();
+	PQExpBuffer newHbaContents = NULL;
+
+	if (hbaLineBuffer == NULL)
+	{
+		log_error("Failed to allocate memory");
+		return false;
+	}
+
+	if (!pghba_append_rule_to_buffer(hbaLineBuffer,
+									 ssl,
+									 databaseType,
+									 database,
+									 username,
+									 host,
+									 authenticationScheme))
+	{
+		/* errors have already been logged */
```

Comment:
Leaks `hbaLineBuffer`, same for all the early returns below.

Reply:
Fixed, thanks.